### PR TITLE
Add pytorch yolo handler

### PR DESCRIPTION
**Pull request by Flova**
## Proposed changes
* Adds a YOLO handler that uses the PyTorch-YOLOv3 repo. This adds GPU support (including yolo v4)

## Necessary checks
- [x] Run `catkin build`
- [x] Test on your machine
- [x] Put the PR on our Project board

Is not a priority to be merged yet, because pytorch yolo doesn't support Yolov4 yet.

### CODE REVIEW

File: bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py [hunk 1]
```diff
@@ -459,6 +462,39 @@ def predict(self):
                     self._candidates[class_name].append(c)
 
 
+class YoloHandlerPytorchYolo(YoloHandler):
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py:465]: Don't like the name that much... YoloHandlerPytorch might be enough.

File: bitbots_vision/src/bitbots_vision/vision_modules/candidate.py [hunk 2]
```diff
@@ -25,6 +25,7 @@ def __init__(self, x1=0, y1=0, width=0, height=0, rating=None):
         self._height = height
         self._rating = rating
 
+
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/candidate.py:28]: Please remove this line

File: bitbots_vision/src/bitbots_vision/vision_modules/candidate.py [hunk 3]
```diff
@@ -218,6 +219,11 @@ def __str__(self):
         """
         return f"x1,y1: {self.get_upper_left_x()},{self.get_upper_left_y()} | width,height: {self.get_width()},{self.get_height()} | rating: {self._rating}"
 
+    @classmethod
+    def from_x1y1x2y2(cls, x1, y1, x2, y2, rating=None):
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/candidate.py:223]: Lol. That's a function name.

File: bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py [hunk 4]
```diff
@@ -459,6 +462,39 @@ def predict(self):
...
+    def predict(self):
+        if self._candidates is None or not self._caching:
+            self._candidates = defaultdict(list)
+            boxes = detect.detect_image(self.model, cv2.cvtColor(self._image, cv2.COLOR_BGR2RGB),
+                                        conf_thres=self._confidence_threshold,
+                                        nms_thres=self._nms_threshold)
+            for box in boxes:
+                # x1, y1, x2, y2, confidence, class
+                c = Candidate.from_x1y1x2y2(*box[0:4].astype(int), box[4].astype(float))
+                classname = self._class_names[int(box[5])]
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py:494]: Not sure whether this step is necessary as the variable names are sufficient in my opinion. But that's just a minor remark.

File: bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py [hunk 5]
```diff
@@ -459,6 +462,39 @@ def predict(self):
...
+
+    def predict(self):
+        if self._candidates is None or not self._caching:
+            self._candidates = defaultdict(list)
+            boxes = detect.detect_image(self.model, cv2.cvtColor(self._image, cv2.COLOR_BGR2RGB),
+                                        conf_thres=self._confidence_threshold,
+                                        nms_thres=self._nms_threshold)
+            for box in boxes:
+                # x1, y1, x2, y2, confidence, class
+                c = Candidate.from_x1y1x2y2(*box[0:4].astype(int), box[4].astype(float))
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py:493]: this line is sweet.

**Review by NFiedler — APPROVED** [review]
LGTM, minor remarks, not tested

File: bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py (hunk 4 above)
Comment (Flova) [reply bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py:494]: What do you mean? The comment?

File: bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py [hunk 6]
```diff
@@ -459,6 +462,39 @@ def predict(self):
...
+                classname = self._class_names[int(box[5])]
+                self._candidates[classname].append(c)
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py:495]:
```suggestion
                self._candidates[self._class_names[int(box[5])]].append(c)
```

File: bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py (hunk 4 above)
Comment (NFiedler) [reply bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py:494]: see suggestion. but really not that important

File: bitbots_vision/src/bitbots_vision/vision_modules/candidate.py (hunk 2 above)
Comment (NFiedler) [reply bitbots_vision/src/bitbots_vision/vision_modules/candidate.py:28]:
```suggestion
```

File: bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py [hunk 7]
```diff
@@ -459,6 +462,38 @@ def predict(self):
...
+class YoloHandlerPytorchYolo(YoloHandler):
+    """
+    Using Pytorch to get YOLO predictions
+    """
+    def __init__(self, config, model_path):
+        """
+        Initialization of PytorchYolo
+
+        :param config:
+        :param model_path:
```
Comment (jaagut) [inline bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py:474]:
```suggestion
        :param config: vision config dict
        :param model_path: path to the yolo model
```

File: bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py [hunk 8]
```diff
@@ -459,6 +462,38 @@ def predict(self):
...
+        """
+        weightpath = os.path.join(model_path, "yolo_weights.weights")
+        configpath = os.path.join(model_path, "config.cfg")
```
Comment (jaagut) [inline bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py:477]: These magic values should be moved somehow or be documented

Comment (johagge) [reply bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py:477]: Do you want more documentation on what a yolo_weights.weights/config.cfg file is?

**Review by jaagut — APPROVED** [review]

**Pull request merged**